chartbeat/chartbeat.go

```
@@ -0,0 +1,68 @@
+package chartbeat
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"regexp"
+	"strings"
+	"time"
+)
+
+var (
+	reg, _ = regexp.Compile("[^0-9]+")
+)
+
+type pages struct {
+	TopPages []TopPage `json:"pages"`
+}
+
+//TopPage maos a top page entry
+type TopPage struct {
+	Path string `json:"path"`
+}
+
+//Repo describe the methods to deal with chartbeat API
+type Repo interface {
+	LoadTopPages(section string, limit int) []string
+}
+
+//RepoImpl implements ChartbeatRepo
+type RepoImpl struct {
+	AuthToken string
+	BaseURL   string
+	Host      string
+}
+
+//LoadTopPages from chartbeat based on section and limit
+func (repo RepoImpl) LoadTopPages(section string, limit int) ([]string, error) {
+	timeout := time.Duration(5 * time.Second)
+	httpClient := http.Client{
+		Timeout: timeout,
+	}
+	url := fmt.Sprintf("%s/?apikey=%s&host=%s&limit=%v&section=%s", repo.BaseURL, repo.AuthToken, repo.Host, limit, section)
+	resp, err := httpClient.Get(url)
+	if err == nil {
```

usually u check the error existence and fail early instead of guarding the happy path